### PR TITLE
Added feature importance methods based on cluster differences

### PR DESCRIPTION
**Pull request by x-tabdeveloping**
I already started working on a fighting-words based feature importance estimation method here: #75 
Since then I have found a better method for measuring semantic difference between clusters based on linear classifiers.
I decided to use LinearDiscrimantAnalysis since it is orders of magnitudes faster than other methods, and it takes forever to estimate feature importance for each of the levels of the hierarchy otherwise when reducing the number of topics.

I am also planning to add supervised topic modelling based on these feature importance methods in the near future.

### CODE REVIEW

File: docs/clustering.md [hunk 1]
```diff
@@ -305,6 +290,50 @@ model = ClusteringTopicModel().fit_dynamic(corpus, timestamps=ts, bins=10)
 model.print_topics_over_time()
 ```
 
+## Semi-supervised Topic Modeling
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply docs/clustering.md:293]: Well, I put it here, because other models can't really do it yet.

File: turftopic/models/cluster.py [hunk 2]
```diff
@@ -96,9 +109,19 @@ def build_tsne(*args, **kwargs):
...
+                super().__init__(*args, **kwargs)
+
+            def fit_transform(self, X: np.ndarray, y=None):
+                return super().fit(X)
+
+            def fit(self, X: np.ndarray, y=None):
+                self.fit_transform(X, y)
+                return self
+
+        return OpenTSNEWrapper(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply turftopic/models/cluster.py:123]: Can you explain?

File: turftopic/models/cluster.py [hunk 3]
```diff
@@ -442,8 +474,12 @@ def fit_predict(
             self.doc_term_matrix = self.vectorizer.fit_transform(raw_documents)
             console.log("Term extraction done.")
             status.update("Reducing Dimensionality")
+            # If y is specified, we pass it to the dimensionality
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply turftopic/models/cluster.py:477]: Well, yes, because the scikit-learn API has a `y` argument in every `fit()` method, regardless of whether it gets used or not. So passing `None` to the method is equivalent to having done nothing.

File: turftopic/models/decomp.py [hunk 4]
```diff
@@ -190,7 +194,11 @@ def fit_transform_multimodal(
                 console.log("Documents encoded.")
             self.embeddings = self.multimodal_embeddings["document_embeddings"]
             status.update("Decomposing embeddings")
-            doc_topic = self.decomposition.fit_transform(self.embeddings)
+            if isinstance(self.decomposition, FastICA) and (y is not None):
+                warnings.warn(
+                    "Supervisory signal is specified but decomposition method is FastICA. y will be ignored. Use a metric learning method for supervised S^3."
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply turftopic/models/decomp.py:199]: Which part is unclear? Can you explain?

File: docs/clustering.md [hunk 5]
```diff
@@ -107,6 +107,25 @@ Turftopic is entirely clustering-model agnostic, and as such, any type of model
...
+    - Can capture more particular word use.
+    - Usually cover the topics' content better.
+2. **Semantic methods**, which estimate term importance using the semantic space of the model:
+    - They typically produce cleaner and more specific topics.
+    - Can be used in a multilingual context.
+    - Generally less sensitive to stop- and junk words.
+
+| Importance method | Type | Description | Advantages |
+| - | - | - | - |
+| `linear` **(NEW)** | Semantic | Project words onto the parameter vectors of a linear classifier (LDA). | Topic differences are measured in embedding space and are determined by predictive power, and are therefore accurate and clean. |
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply docs/clustering.md:124]: That is correct, yes

File: turftopic/models/cluster.py [hunk 6]
```diff
@@ -422,7 +453,8 @@ def fit_predict(
         raw_documents: iterable of str
             Documents to fit the model on.
         y: None
-            Ignored, exists for sklearn compatibility.
+            Originally ignored, in case of a dimensionality reduction that can utilize labels,
+            you can pass labels to the model to inform the clustering process.
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply turftopic/models/cluster.py:457]: I rephrased to make it clearer

File: docs/clustering.md [hunk 7]
```diff
@@ -120,43 +139,6 @@ Multiple methods are available in Turftopic for estimating words'/phrases' impor
         # or
         model = ClusteringTopicModel(feature_importance="c-tf-idf")
         ```
-        !!! failure inline end "Weaknesses"
```
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply docs/clustering.md:123]: I added them back

File: turftopic/models/decomp.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply turftopic/models/decomp.py:203]: Hmm better, but would update to: 

```suggestion
                    "Supervisory signal (`y`) is specified, but the decomposition method is FastICA, which does not support a supervisory signal. `y` will be ignored. To use supervised s^3, ____"
```
Where for ___ I would either put in a recommendation 

> use decomposition=valid-value

Or refer to the docs on where you can find an introduction or a list of valid values.

File: turftopic/models/cluster.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (KennethEnevoldsen) [reply turftopic/models/cluster.py]: Better, but 

> In case the dimensionality reduction that can utilize labels, then you can pass labels to the model to inform the clustering process. If the dimensionality reduction does not support labels, this will be ignored. 

Optionally add:
>To see more on supervised clustering see ...

File: turftopic/models/decomp.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (x-tabdeveloping) [reply turftopic/models/decomp.py:203]: I will update the documentation on this once the feature becomes more mature. I do not know terribly much about what works and what doesn't just yet, so I will need to mess around with this more.
